atlasdb-commons/src/main/java/com/palantir/common/proxy/ReplaceIfExceptionMatchingProxy.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.common.proxy;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class ReplaceIfExceptionMatchingProxy<T> implements InvocationHandler {
+    private static final Logger log = LoggerFactory.getLogger(ReplaceIfExceptionMatchingProxy.class);
+
+    private final Supplier<T> delegateFactory;
+    private final Predicate<Throwable> shouldReplace;
+    private volatile T delegate;
+
+    private ReplaceIfExceptionMatchingProxy(Supplier<T> delegateFactory, Predicate<Throwable> shouldReplace) {
+        this.delegateFactory = delegateFactory;
+        this.delegate = delegateFactory.get();
+        this.shouldReplace = shouldReplace;
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+        try {
+            return method.invoke(delegate, args);
+        } catch (InvocationTargetException e) {
+            Throwable cause = e.getCause();
+            replaceIfNecessary(cause);
+            throw cause;
+        } catch (IllegalAccessException e) {
+            throw new IllegalStateException(e);
```

Comment:
Is handling this separately necessary?

Reply:
We probably don't want to throw the cause in this case (seems like a bug in our reflection so we probably want to see it)

Reply:
i fixed logging

Reply:
I checked in other places in this repo, and I don't really see it being handled so I'm curious.

Reply:
yeah, this was just an artifact of me removing 'throws Throwable', handling the cases and then realizing i needed to rethrow the cause

Reply:
Yeah, this is what's failing the build